server/rest/src/main/java/org/infinispan/rest/authentication/impl/BasicAuthenticator.java

```diff
@@ -27,24 +30,23 @@ public BasicAuthenticator(SecurityDomain domain, String realm) {
    }
 
    @Override
-   public void challenge(NettyRestRequest request, ChannelHandlerContext ctx) throws AuthenticationException {
+   public CompletionStage<RestResponse> challenge(RestRequest request, ChannelHandlerContext ctx) {
       String auth = request.getAuthorizationHeader();
-      if (auth == null) throw new AuthenticationException(Optional.of(authenticateHeader));
-      if (auth.length() > 5) {
+      if (auth != null && auth.length() > 5) {
          String type = auth.substring(0, 5);
          type = type.toLowerCase();
          if ("basic".equals(type)) {
             String cookie = auth.substring(6);
             cookie = new String(Base64.getDecoder().decode(cookie.getBytes()));
             String[] split = cookie.split(":");
             try {
-               request.setPrincipal(domain.authenticate(split[0], split[1]));
+               request.setSubject(domain.authenticate(split[0], split[1]));
```

Comment:
It seems the ```challenge()``` method will still block on ```domain.authenticate()```?
This guy takes a large chunk of the REST request processing time I'm afraid.

Reply:
Right, I'll wrap that

Reply:
It's now wrapped in a CompletableFuture.supplyAsync